faunadb-java-dsl/src/main/java/com/faunadb/client/types/Constructors.java

```diff
@@ -0,0 +1,211 @@
+
+                return new ConstructorCodec(constructor);
+            }
+        }
+
+        return null;
+    }
+
+    private static Codec<Object> getDefaultConstructorCodec(Class<?> clazz) {
+        try {
+            return new DefaultConstructorCodec(clazz.getDeclaredConstructor());
+        } catch (NoSuchMethodException ex) {
+            return null;
+        }
+    }
+
+    static private abstract class AbstractCreator implements Codec<Object> {
+        final private Class<?> rawClass;
+        final private String[] parameterNames;
+        final private SimpleType[] parameterTypes;
+        final private Property[] writeProperties;
+
+        AbstractCreator(Constructor<?> constructor) {
+            this.rawClass = constructor.getDeclaringClass();
+            this.parameterNames = getParameterNames(constructor.getParameterAnnotations());
+            this.parameterTypes = getParameterTypes(constructor.getGenericParameterTypes());
+            this.writeProperties = Properties.getWriteProperties(constructor.getDeclaringClass());
+        }
+
+        AbstractCreator(Method method) {
+            this.rawClass = method.getDeclaringClass();
+            this.parameterNames = getParameterNames(method.getParameterAnnotations());
+            this.parameterTypes = getParameterTypes(method.getGenericParameterTypes());
+            this.writeProperties = Properties.getWriteProperties(method.getDeclaringClass());
+        }
+
+        private SimpleType[] getParameterTypes(Type[] genericParameterTypes) {
+            SimpleType[] parameterTypes = new SimpleType[genericParameterTypes.length];
+
+            for (int i = 0; i < genericParameterTypes.length; i++) {
+                parameterTypes[i] = Types.of(genericParameterTypes[i]);
+            }
+
+            return parameterTypes;
+        }
+
+        private String[] getParameterNames(Annotation[][] parameterAnnotations) {
+            String[] parameterNames = new String[parameterAnnotations.length];
+
+            for (int i = 0; i < parameterAnnotations.length; i++) {
+                Annotation[] annotations = parameterAnnotations[i];
+
+                for (Annotation annotation : annotations) {
+                    if (annotation.annotationType() == FaunaField.class) {
+                        parameterNames[i] = ((FaunaField) annotation).value();
+                    }
+                }
+
+                if (parameterNames[i] == null)
+                    throw new FaunaException(format("All constructor/creator arguments must be annotated with @%s", FaunaField.class.getSimpleName()));
+            }
+
+            return parameterNames;
+        }
+
+        @Override
+        public Result<Object> apply(Value value) {
+            try {
+                ObjectV object = (ObjectV) value;
+                Object instance = newInstance(buildArguments(object));
+
+                for (Property property : writeProperties) {
```

Comment:
I think we should filter out the properties already used by the constructor so we avoid calling a setter unnecessarily.